File: go/id_generator.go
package messages

import (
	"strconv"

	"github.com/google/uuid"
)

type IdGenerator interface {
	newId() func() string
}

type Incrementing struct {
	next int
}

func (self *Incrementing) NewId() string {
	result := strconv.Itoa(self.next)
	self.next++
	return result
}

type UUID struct {
	next int
}

func (i UUID) NewId() string {
	return uuid.New().String()
}

File: dart/lib/src/id_generator.dart
/// Generates a new unique identifier string.
///
/// Implement this typedef in your application (for example by delegating to a
/// UUID library you already depend on). The package only ships
/// [incrementingIdGenerator] for tests.
typedef IdGenerator = String Function();

/// Returns an [IdGenerator] that produces incrementing decimal strings
/// (`"0"`, `"1"`, …).
///
/// Intended for tests; not safe for production use across processes.
IdGenerator incrementingIdGenerator() {
  var next = 0;
  return () => '${next++}';
}
